Automake and libtool

inconstante opened the pull request:
This patch introduces the use of automake and libtool.

Several files have been removed, most notably Makefile.in, since this file is now automatically generated by automake, based on a collection of Makefile.am files.

On the one hand, the Makefile.am file in the src directory uses automake and libtool directives in order to generate src/Makefile.in. On the other hand, the Makefile.am file in the test directory contains several make rules, written by hand, so that the tests (make check) mimic the current tests. It is out of the scope of this patch set to modify the tests.

Also notable, is the removal of the configure script from the version control. This follows the recommendation on https://www.sourceware.org/autobook/autobook/autobook_43.html.

From a fresh checkout, a developer would like to use the following commands:

$ ./bootstrap
$ ./configure
$ make
$ make check
$ make install

The distribution rules, which create a distributable tar ball with the configure script, are also useful for developers:

$ make dist
$ make distcheck

Warning: The dist and distcheck rules currently depend on the all rule, thus a developer must run make before make dist or make distcheck. This is due to the handmade test/Makefile.am file and could be improved.

inconstante committed:
Converts project to use automake and libtool

This commit introduces the use of automake and libtool on libauxv.

A new set of Makefile.am files is now provided for automake and libtool.

A bootstrap script is also provided. Therefore, automatically generated files,
such as configure, should never be included in the version control system.

The source files: src/auxv.c and src/lsauxv.c have been modified because the
header files have been moved into their distribution directory: include/auxv/

inconstante committed:
Ignores autotools log and cache files

inconstante committed:
Re-enables man pages installation

When converting this project to automake and libtool, the installation of the
man pages had been temporarily disabled. This commit re-enables it.

inconstante committed:
Removes options automatically provided by libtool

The options --enable-shared and --enable-static, as well as their disable
counterparts, are automatically provided by libtool. Therefore there is no need
to explicitly add them in configure.ac.

inconstante committed:
Prepares directory structure for automake

NOTE: This commit only moves files. It does not modify its contents!

The following files have been deleted, because they will be automatically
generated by autoconf and automake:
  Makefile.in
  config.h
  config.in
  configure

The file autoscan.log has also been deleted and added to the .gitignore list.

The source and header files, man pages, and tests have been moved to their own
subdirectories, as recommended by the autotools manual.

inconstante committed:
Re-enables the use of the --with-headers switch

When converting the project to automake and libtool, the --with-headers switch
had been temporarily disabled. This commit re-enables it.

inconstante committed:
Fixes make distcheck error

inconstante committed:
Re-enables tests with make check

When converting the project to automake and libtool, the tests had been
temporarily disabled. This commit adds a new Makefile.am file, which re-enables
tests on libauxv.

tuliom commented on bootstrap:
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+autoscan
```
Why can't you use autoreconf here?

inconstante replied:
That would be because I removed all files automatically generated by autotools from the version control, more specifically those created by the first execution of both: 'libtoolize' (ltmain.sh) and 'automake --add-missing --force-missing' (ar-lib, compile, config.guess, config.sub, depcomp, install-sh, missing, test-driver). Running autoreconf will fail, unless you run these bootstrap commands in that particular order.

I did not know that users should be able to build the project directly from the repository without having to install GNU auto tools, thus I removed all automatically generated files. That's something that I have already seen before in some projects, e.g.: wayland and weston.

I was hoping that we could use 'make dist' in order to create the distributable tar ball, which does contain configure, Makefile.in, etc.

tuliom replied:
> Running autoreconf will fail, unless you run these bootstrap commands in that particular order.

'autoreconf -f -i' doesn't fail here. Or am I missing something?

> I was hoping that we could use 'make dist' in order to create the distributable tar ball, which does contain configure, Makefile.in, etc.

No, we can't adopt 'make dist' only. Many of the enterprise distributions have old versions of the GNU Autotools which usually lack features we use during development. If we want to be able to build every commit ID on these distributions, we have to provide all these files in the repository and enable Autotools maintainer mode.

tuliom commented on .gitignore:
```diff
@@ -1,2 +1,20 @@
+config.guess
+config.h.in
+config.sub
+configure
```
We need configure.
Users must be able to build the project directly from the repository without having to install GNU auto tools.
That's how it used to work.

tuliom commented on .gitignore:
```diff
@@ -1,2 +1,20 @@
+install-sh
+ltmain.sh
+m4/
+man/Makefile.in
```
We'll need all these Makefile.in files in order to let configure work.

inconstante committed:
Re-inserts autotools generated files

The files added by this commit are automatically generated by autoconf,
automake, and libtool. They have been (re-)added to the version control, since
this project is expected to be buildable from fresh clones of the git
repository in machines that do not have the autotools installed.

inconstante committed:
Fixes dependency on libpthread

The flag -lpthread was being passed to the linking step of the lsauxv binary
generation, however such binary does not use any of symbols provided by
pthread, as pointed out by dpkg-shlibdeps:

  dpkg-shlibdeps: warning: package could avoid a useless dependency if
  debian/libauxv/usr/bin/lsauxv was not linked against libpthread.so.0 (it uses
  none of the library's symbols).

On the other hand, the same flag [-lpthread] was not being passed to the
linking step of the libauxv.so generation. However, it should have, since
libauxv uses the symbol pthread_once, as pointed out by dpkg-shlibdeps:

  dpkg-shlibdeps: warning: symbol pthread_once used by
  debian/libauxv/usr/lib/powerpc64le-linux-gnu/libauxv-1.0.0.so found in none
  of the libraries.

tuliom commented on bootstrap:
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+autoreconf -f -i
```
If autoreconf fails, this script will return 0.
Could you add a "set -e" at its beginning, please?
